[FSDP] Implement reshard_flatten_tensor

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75192
* #75537

Implement reshard_flatten_tensor() to allow FSDP to reshard the flatten tensor
from equally sharding (chunk) to any other one-dimensional sharding.

Differential Revision: [D35361572](https://our.internmc.facebook.com/intern/diff/D35361572/)

fegin committed:
[FSDP] Implement reshard_flatten_tensor

Implement reshard_flatten_tensor() to allow FSDP to reshard the flatten tensor
from equally sharding (chunk) to any other one-dimensional sharding.

Differential Revision: [D35361572](https://our.internmc.facebook.com/intern/diff/D35361572/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/shard_utils.py:
```diff
@@ -0,0 +1,142 @@
+    return input_split_sizes, output_split_sizes
+
+
+def reshard_flatten_tensor(
```
It looks like this API isn't tested at the moment?

fegin committed:
Update on "[FSDP] Implement reshard_flatten_tensor"


Implement reshard_flatten_tensor() to allow FSDP to reshard the flatten tensor
from equally sharding (chunk) to any other one-dimensional sharding.

Differential Revision: [D35361572](https://our.internmc.facebook.com/intern/diff/D35361572/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement reshard_flatten_tensor"


Implement reshard_flatten_tensor() to allow FSDP to reshard the flatten tensor
from equally sharding (chunk) to any other one-dimensional sharding.

Differential Revision: [D35361572](https://our.internmc.facebook.com/intern/diff/D35361572/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement reshard_flatten_tensor"


Implement reshard_flatten_tensor() to allow FSDP to reshard the flatten tensor
from equally sharding (chunk) to any other one-dimensional sharding.

Differential Revision: [D35361572](https://our.internmc.facebook.com/intern/diff/D35361572/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement reshard_flatten_tensor"


Implement reshard_flatten_tensor() to allow FSDP to reshard the flatten tensor
from equally sharding (chunk) to any other one-dimensional sharding.

Differential Revision: [D35361572](https://our.internmc.facebook.com/intern/diff/D35361572/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement reshard_flatten_tensor"


Implement reshard_flatten_tensor() to allow FSDP to reshard the flatten tensor
from equally sharding (chunk) to any other one-dimensional sharding.

Differential Revision: [D35361572](https://our.internmc.facebook.com/intern/diff/D35361572/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Implement reshard_flatten_tensor"


Implement reshard_flatten_tensor() to allow FSDP to reshard the flatten tensor
from equally sharding (chunk) to any other one-dimensional sharding.

Differential Revision: [D35361572](https://our.internmc.facebook.com/intern/diff/D35361572/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/shard_utils.py:
```diff
@@ -0,0 +1,144 @@
+        my_rank (int): the rank for this trainer.
+
+    Returns:
+        The local shard for the new ShardedTensor.
```
Is there a reason we're returning a torch.Tensor which is the local shard instead of a ShardedTensor instance?

rohan-varma commented on torch/distributed/fsdp/shard_utils.py:
```diff
@@ -0,0 +1,144 @@
+) -> torch.Tensor:
+    """
+    Resharded a sharded flatten tensor, this is used by FSDP to do sharded
+    state_dict. But the functionaility is not supported by ShardedTensor.
```
Should we eventually generalize + upstream `reshard_flatten_tensor` if it is useful more generally in sharding package? cc @pritamdamania87 @wanchaol

fegin replied:
I discussed with @fduwjj , if reshard_flatten_tensor is general, we can upstream it to the sharding package. Right now, there are some gaps.

rohan-varma commented on torch/distributed/fsdp/shard_utils.py:
```diff
@@ -0,0 +1,144 @@
+
+    This API uses the ChunkShardingSpec and EnumerableShardingSpec from
+    torch.distributed.sharding_spec but ignores the placement field in
+    ChunkShardingSpec, as the placement requires the callees understand the
```
Could this mean that the resulting tensor would not be placed on a device specified by `ChunkShardingSpec`?

fegin replied:
Yes, this is the gap to upstream this API to the sharding package. Making FSDP knowing how to create the remote device is not general. As a result, this API currently ignore the device field.

rohan-varma commented on torch/distributed/fsdp/shard_utils.py:
```diff
@@ -0,0 +1,144 @@
+    ShardingSpec,
+)
+from torch.distributed.nn.functional import (
+    all_to_all_single,
```
I'm not sure if we actually formally support comm. APIs in dist.nn.functional, as from previous discussions I remember they had/have a couple of issues (@cbalioglu @mrshenli probably know better). In general, the APIs in `distributed_c10d` are battle-tested and well-supported, so could we consider using a2a from there, unless there is a specific need to have autograd know about this a2a call?

rohan-varma reviewed (approved):
LGTM! One concern around usage of all_to_all API which may require inputs from other folks.

fegin committed:
Update on "[FSDP] Implement reshard_flatten_tensor"


Implement reshard_flatten_tensor() to allow FSDP to reshard the flatten tensor
from equally sharding (chunk) to any other one-dimensional sharding.

Differential Revision: [D35361572](https://our.internmc.facebook.com/intern/diff/D35361572/)

[ghstack-poisoned]
